raster/r.kappa/calc_metrics.c

```diff
@@ -0,0 +1,203 @@
+#include <stdlib.h>
+#include <grass/gis.h>
+#include <grass/glocale.h>
+#include "kappa.h"
+#include "local_proto.h"
+
+static int longcomp(const void *aa, const void *bb);
+static int collapse(long *l, int n);
+
+
+void calc_metrics(void)
+{
+    int i, j, k;
+    long *clst;
+    int ncat1, ncat2;
+    int cndx;
+    double *pi, *pj, *pii;
+    double p0 = 0, pC = 0;
+    double inter1 = 0, inter2 = 0;
+    int a_i, b_i;
+
+    /* get the cat lists */
+    rlst = (long *)G_calloc(nstats * 2, sizeof(long));
+    clst = (long *)G_calloc(nstats, sizeof(long));
+    for (i = 0; i < nstats; i++) {
+        rlst[i] = Gstats[i].cats[0];
+        clst[i] = Gstats[i].cats[1];
+    }
+
+    /* sort the cat lists */
+    qsort(rlst, nstats, sizeof(long), longcomp);
+    qsort(clst, nstats, sizeof(long), longcomp);
+
+    /* remove repeated cats */
+    ncat1 = collapse(rlst, nstats);
+    ncat2 = collapse(clst, nstats);
+
+    /* copy clst to the end of rlst, remove repeated cats, and free unused memory */
+    for (i = 0; i < ncat2; i++)
+        rlst[ncat1 + i] = clst[i];
+    qsort(rlst, ncat1 + ncat2, sizeof(long), longcomp);
+    ncat = collapse(rlst, ncat1 + ncat2);
+    rlst = (long *)G_realloc(rlst, ncat * sizeof(long));
+    G_free(clst);
+
+    /* allocate matrix and fill in with cats' value */
+    metrics = (METRICS *) G_malloc(sizeof(METRICS));
+    metrics->matrix = (long *)G_malloc(ncat * ncat * sizeof(long));
```

Comment:
```suggestion
    metrics->matrix = (long *)G_malloc((size_t)ncat * ncat * sizeof(long));
```
should be enough.

Reply:
Thank you for your hint. I am not good at C :-(